### DISCUSSION TITLE
Can't correctly setup an EventSource.

### DISCUSSION THREAD

**Discussion by ENBYSS** [nuxt/nuxt#16046] [Questions]
So, I want to implement an API that returns an EventSource which the client can then subscribe to in order to receive updates - however I've encountered numerous roadblocks.

# Res and Req
First, I tried using the **res** and **req** attributes:

```ts
export default defineEventHandler(event => {
        event.node.res.writeHead(200, {
		'Content-Type': 'text/event-stream',
		'Connection': 'keep-alive',
		'Cache-Control': 'no-cache'
	});
	
	const connection = {
		id: crypto.randomUUID(),
		res: event.node.res
	};
	
	connections.push(connection);

	console.log(`Opened connection: ${connection.id}`);

	const closer = (badge: string) => {
		console.log(`${badge} Closing connection: ${connection.id}`);
		connections = connections.filter(client => client.id !== connection.id);
	}

	event.node.req.on('close', () => closer("[CLOSE]"));
	event.node.req.on('end', () => closer("[ END ]"));
	event.node.req.on('error', () => closer("[ERROR]"));
	event.node.req.on('pause', () => closer("[PAUSE]"));
});
```

This is the most functional of the lot I'll be showing - but it's not perfect. The client can subscribe to this, however the server throws the following error:
```
[nitro] [dev] [unhandledRejection] Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client                                                                                                         16:11:09
    at new NodeError (node:internal/errors:393:5)
    at ServerResponse.setHeader (node:_http_outgoing:644:11)
    at Object.errorhandler [as onError] (file:///C:/Users/ENBYSS/Desktop/the-abyss/enbyss.com/my-blog/.nuxt/dev/index.mjs:469:20)
    at Server.toNodeHandle (file:///C:/Users/ENBYSS/Desktop/the-abyss/enbyss.com/my-blog/node_modules/h3/dist/index.mjs:752:27)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  code: 'ERR_HTTP_HEADERS_SENT'
```

I wouldn't mind this, however the more impactful downside is that any **.on** methods fail to work. As a result, connections just keep increasing even as clients disconnect. I just get multiple `Opened connection` logs.

# setHeader instead of writeHead
In order to try and circumvent this, I then tried to use `setHeader` instead of `writeHead`.

```ts
export default defineEventHandler(event => {
	event.node.res.statusCode = 200;
	event.node.res.setHeader("Content-Type", "text/event-stream");
	event.node.res.setHeader("Connection", "keep-alive");
	event.node.res.setHeader("Cache-Control", "no-cache");
	
	const connection = {
		id: crypto.randomUUID(),
		res: event.node.res
	};
	
	connections.push(connection);

	console.log(`Opened connection: ${connection.id}`);

	const closer = (badge: string) => {
		console.log(`${badge} Closing connection: ${connection.id}`);
		connections = connections.filter(client => client.id !== connection.id);
	}

	event.node.req.on('close', () => closer("[CLOSE]"));
	event.node.req.on('end', () => closer("[ END ]"));
	event.node.req.on('error', () => closer("[ERROR]"));
	event.node.req.on('pause', () => closer("[PAUSE]"));
});
```

Now, the good news is that the `ERR_HTTP_HEADERS_SENT` error disappears. The bad news is that the client errors out with a **404** - even though the server accurately logs the connection. Infact the full log ends up being this:

```
Opened connection: 370f1d42-139e-44b4-88d2-cad43bbff97f                                                                                                                                                                      
[ END ] Closing connection: 370f1d42-139e-44b4-88d2-cad43bbff97f                                                                                                                                                               
[CLOSE] Closing connection: 370f1d42-139e-44b4-88d2-cad43bbff97f   
```

Showing that it *did* register a connection, but it then instantly dropped it instead of holding on. My theory is that this is because the handler returns nothing... which is on purpose since I want to *write* to the connection and not end. If I try to return an EventSource compatible string like this:

```ts
return `id: ${con.id}\ndata: ${JSON.stringify(events)}\n\n`;
```

The EventSource finds the data, then immediately gets disconnected because the connection dropped.

# Setting headers with H3
After all this, I thought that maybe the solution was trying to just use H3 utilities.

```ts
export default defineEventHandler(event => {
        event.node.res.statusCode = 200;
	setResponseHeaders(event, {
		'Content-Type': 'text/event-stream',
		'Connection': 'keep-alive',
		'Cache-Control': 'no-cache'
	})
	
	const connection = {
		id: crypto.randomUUID(),
		res: event.node.res
	};
	
	connections.push(connection);

	console.log(`Opened connection: ${connection.id}`);

	const closer = (badge: string) => {
		console.log(`${badge} Closing connection: ${connection.id}`);
		connections = connections.filter(client => client.id !== connection.id);
	}

	event.node.req.on('close', () => closer("[CLOSE]"));
	event.node.req.on('end', () => closer("[ END ]"));
	event.node.req.on('error', () => closer("[ERROR]"));
	event.node.req.on('pause', () => closer("[PAUSE]"));
});
```

This ended up having the same effect as using `setHeader` - the connection just gets dropped immediately. I tried setting the response code using H3, but I couldn't find a utility for that - the one in the GitHub docs doesn't work, and the other docs don't mention a way from what I could tell.

Now, using the same code *without* h3 in a separate node server ends up working perfectly - although this is pretty cumbersome and complicates everything more.

In essence, I just want to implement a working `EventSource`. Either by using some h3 utility I'm not aware of, or by somehow bypassing h3 entirely and using express' `res` and `req`. I'm not sure if this is even possible, but I'm at my wits end here...

**Comment by ENBYSS** [comment]
Did some further experimentation.

First, I managed to create the _legacy_ version by using `fromNodeMiddleware` instead of `defineEventListener` - however it still doesn't work and leads to the same problem.

Second, I create a much more minimal example to experiment more quickly.
```ts
export default defineEventHandler(async (e) => {
    //Returning headers to the client
    e.node.res.writeHead(200, {
        'Cache-Control': 'no-cache',
        Connection: 'keep-alive',
        'Content-Type': 'text/event-stream',
    });

    const sseId = crypto.randomUUID();

    const sendEvent = (data: string) => {
        e.node.res.write(`id: ${sseId}\n`);
        e.node.res.write(`data: ${data}\n\n`);
      };

    console.log("Opened connection:", sseId);
    const closer = (badge: string) => {
		console.log(`${badge} Closing connection: ${sseId}`);
	}

    e.node.req.on('close', () => closer("[CLOSE]"));
	e.node.req.on('end', () => closer("[ END ]"));
	e.node.req.on('error', () => closer("[ERROR]"));
	e.node.req.on('pause', () => closer("[PAUSE]"));

    e.node.req.addListener("close", () => closer("[CLOSE]"));
	e.node.req.addListener('end', () => closer("[ END ]"));
	e.node.req.addListener('error', () => closer("[ERROR]"));
	e.node.req.addListener('pause', () => closer("[PAUSE]"));

    let i = 0;
    setInterval(() => {
        const data = {date:new Date(), eventCount:i++};
        sendEvent(JSON.stringify(data));
    }, 1000);

    const ender : () => {} = await new Promise(resolve => setTimeout(() => resolve(() => e.node.res.end()), 10_000));
    sendEvent('[DONE]');
    ender();
});
```

With this, I found out two things:
1. The reason for the header error is because the request is continuing to write _after_ being "ended". In the case listed above the error doesn't show up - meaning that technically logic should be implemented in order to block the event handler until the request is complete.
2. The listeners however _still_ don't work. If a client access the endpoint and closes the browser before the request completes, the events don't trigger. Infact, they only trigger _when_ `ender` is called. It doesn't matter what the client does - the only thing that consistently triggers the listeners is a call to `.end()` the request.

I'm unsure if this is the **intended** functionality - but if so an alternative to get the expected functionality would be nice - be it with h3 or something else. I couldn't manage to integrate `express` and cheat by using that, since it seems that the `serverMiddleware` property that most guides depend on was removed - and without it the code doesn't end up working.

**Comment by ENBYSS** [comment]
Unfortunately the solution I had to go for is separating my event source code into a separate node server. I don't know if I'm missing something or SSE's weren't tested - even googling seems to bring up no actionable solutions. I am admittedly a bit frustrated here, so hopefully I'm missing the obvious or this is a known issue.

For anyone with the same issue, yeah. Hope this helps. If you own a server you just need to deploy a node.js server with express and dispatch event sources that way. If you're serverless... maybe the incomplete solution above can help. Not sure if it works, and if it does it might crash due to a never-ending increase in connections, but if it's locked via auth that might be mitigated.

**Comment by Gringo675** [comment]
ENBYSS, thanks for detailed posts. Sure it helps to save much time.
I’ve played with SSE too, with generally the same result.
It partly works, as long as you deliberately not closing the connection, for example with endless await, like this:
```
const listeners = []
const sendRandom = () => {
    const data = Math.random()
    listeners.forEach(listener => {
        listener.event.node.res.write(`event: random\ndata: ${JSON.stringify(data)}\n\n`)
    })
}
setInterval(sendRandom, 5000)

export default defineEventHandler(async (event) => {
    event.node.res.writeHead(200, {
        'Cache-Control': 'no-cache',
        Connection: 'keep-alive',
        'Content-Type': 'text/event-stream',
    });
    listeners.push(event)
    await new Promise(() => {})
})
```

But yes, event listeners don’t work, so you can’t delete closed connections.

Then I’ve tried streams, like this

```
import { PassThrough } from 'node:stream'
import { sendStream } from 'h3'
const tunnel = new PassThrough();

const writeToTunnel = () => {
    const data = Math.random()
    tunnel.write(`event: random\ndata: ${JSON.stringify(data)}\n\n`)
}
setInterval(writeToTunnel, 5000)

export default defineEventHandler((event) => {
    event.node.res.writeHead(200, {
        'Cache-Control': 'no-cache',
        'Connection': 'keep-alive',
        'Content-Type': 'text/event-stream',
    });

    return sendStream(event, tunnel)
})
```
It works just as the previous code, so we can receive events, but event listeners on server still don't work.

Then I’ve monitored node server connections. And see that after I closed connection on client, on server it goes into ‘Time wait’ state and after a while (about 2 min) closed absolutely. So I think maybe its ok not to manually clear closed connections, as they closed by themselves after a while, and we can go without event listeners. But think this behavior depends on the specific environment. My nuxt app run on node in Windows 10.
It would be good to hear the developers opinion on this topic

P.S. If someone need client code for SSE:
```
<script setup>
let eventSource
const start = () => {
  eventSource = new EventSource('/api/…');

  eventSource.addEventListener('random', event => {
    console.log("New data", event.data);
  });
}
const close = () => {
  eventSource.close()
}
</script>

<template>
  <div class="console">
    <h1>Console</h1>
    <button class="button" @click="start">start</button>
    <button class="button" @click="close">close</button>
  </div>
</template>
```

**Comment by didavid61202** [comment]
Hi @enbyss, @Gringo675, I've found a possible cause for this issue in the 'httpxy' lib (deps of nitro) used by nitro during dev, will create a PR if @pi0 think it's the right way to go 👍
more detail in Atinux's SSE gist comment -> https://gist.github.com/Atinux/05836469acca9649fa2b9e865df898a2
